nemo/collections/nlp/models/nlp_model.py

```diff
@@ -36,8 +46,76 @@ class NLPModel(ModelPT):
+                self.register_artifact(encoder_config_path, encoder_config_src)
+            elif isinstance(self.bert_model, MegatronBertEncoder):
+                config_for_json = OmegaConf.to_container(self.bert_model.config)
+                with open(encoder_config_src, 'w', encoding='utf-8') as f:
+                    f.write(json.dumps(config_for_json, indent=2, sort_keys=True) + '\n')
+                pretrained_model_name = self.bert_model._model_name
+                encoder_config_path = pretrained_model_name + '_encoder_config.json'
+                self.register_artifact(encoder_config_path, encoder_config_src)
+            else:
+                logging.info(
+                    f'Registering BERT model config for {self.bert_model} is not yet supported. Please override this method if needed.'
+                )
+
+    def _setup_tokenizer(self, cfg: DictConfig):
+        tokenizer = get_tokenizer(
+            tokenizer_name=cfg.tokenizer_name,
+            vocab_file=cfg.vocab_file,
+            special_tokens=OmegaConf.to_container(cfg.special_tokens) if cfg.special_tokens else None,
+            tokenizer_model=cfg.tokenizer_model,
+        )
+        self.tokenizer = tokenizer
+        self.register_tokenizer(cfg)
+
+    @rank_zero_only
+    def register_tokenizer(self, cfg: DictConfig):
+        vocab_file_config_path = 'tokenizer.vocab_file'
+        vocab_dict_config_path = 'tokenizer_vocab_dict.json'
+        if self.tokenizer is None:
+            raise ValueError('Instantiate self.tokenizer before registering it.')
+        else:
+            if cfg.vocab_file is not None:
+                self.register_artifact(config_path=vocab_file_config_path, src=cfg.vocab_file)
+            elif isinstance(self.tokenizer, AutoTokenizer):
+                # extract vocab from tokenizer
+                vocab_json_src = os.path.join(NEMO_NLP_TMP, vocab_dict_config_path)
+                vocab_dict = self.tokenizer.tokenizer.vocab
+                with open(vocab_json_src, 'w', encoding='utf-8') as f:
+                    f.write(json.dumps(vocab_dict, indent=2, sort_keys=True) + '\n')
+                self.register_artifact(vocab_dict_config_path, vocab_json_src)
+                # create vocab file
+                vocab_file_src = os.path.join(NEMO_NLP_TMP, vocab_file_config_path)
+                with open(vocab_file_src, 'w', encoding='utf-8') as f:
+                    for key in vocab_dict:
+                        f.write(key + '\n')
+                self.register_artifact(config_path=vocab_file_config_path, src=cfg.vocab_file)
+            else:
+                logging.info(
+                    f'Registering tokenizer vocab for {self.tokenizer} is not yet supported. Please override this method if needed.'
+                )
+            if cfg.tokenizer_model is not None:
+                self.register_artifact(config_path='tokenizer.tokenizer_model', src=cfg.tokenizer_model)
```

Comment:
Ok, I don't get it - what really is "tokenizer model"? Is it a pure "name", part of config section, or sth else?